File: phpoc-web/test/commonplace_push_service_test.mjs
/**
 * commonplace_push_service_test.mjs — CommonplacePushService test suite (Phase 2 RED).
 *
 * Group P (9) from docs/planning/COMMONPLACE_BOOK_SYNC_WEB_PHASE1.md.
 * Pure Node tests on the KeyedMockCrypto + FakeSyncTransport harness.
 *
 * RED: targets the not-yet-created `src/commonplace/commonplace_push_service.js`
 * (CommonplacePushService). Every assertion here fails until that module exists
 * and its pushAll/pushBlocks methods pass.
 *
 * Run: node test/commonplace_push_service_test.mjs
 */

import { MemoryBackend } from '../src/sync/storage.js';
import { TestHelpers } from './test_helpers.mjs';
import { CommonplaceChain } from '../src/commonplace/commonplace_chain.js';
import { CommonplacePushService } from '../src/commonplace/commonplace_push_service.js';
import {
  syncTestMkHex,
  makeCrypto,
  jsonSortNoSpaces,
  FakeSyncTransport,
  buildChain,
  commonplaceBlockPath,
  COMMONPLACE_HASH_INDEX,
  decodeStoredBlock,
  readHashIndex,
} from './commonplace_sync_test_support.mjs';

const t = new TestHelpers();

async function test(name, fn) {
  try {
    await fn();
  } catch (e) {
    t.assert(false, `${name} — ${e.message}`);
  }
}

// ── CPSW-P1 ────────────────────────────────────────────────────────
await test('CPSW-P1: pushAll pushes every block to commonplace/blocks/NNNNNN.json', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  const result = await svc.pushAll();

  t.assertEq(result.pushed, 2, 'P1: two blocks pushed');
  t.assert(Array.isArray(result.failedBlocks) && result.failedBlocks.length === 0, 'P1: no failed blocks');
  t.assert(transport.store.has(commonplaceBlockPath(0)), 'P1: 6-digit index 0 present');
  t.assert(transport.store.has(commonplaceBlockPath(1)), 'P1: 6-digit index 1 present');

  const genesis = JSON.parse(decodeStoredBlock(transport, 0, crypto, syncTestMkHex));
  t.assertEq(genesis.type, 'commonplace_genesis', 'P1: genesis type');
  const day = JSON.parse(decodeStoredBlock(transport, 1, crypto, syncTestMkHex));
  t.assertEq(day.type, 'commonplace', 'P1: day type');
});

// ── CPSW-P2 ────────────────────────────────────────────────────────
await test('CPSW-P2: pushAll pushes plaintext hash_index of block hashes in order', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  await svc.pushAll();

  const index = readHashIndex(transport);
  const blocks = await chain.readAll();
  t.assertEq(index.length, 2, 'P2: hash index length 2');
  t.assertEq(index[0], chain.getBlockHashFor(blocks[0]), 'P2: genesis hash');
  t.assertEq(index[1], chain.getBlockHashFor(blocks[1]), 'P2: day hash');
});

// ── CPSW-P3 ────────────────────────────────────────────────────────
await test('CPSW-P3: pushed block payload is sorted-keys compact JSON obfuscated with MK', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  await svc.pushAll();

  const blocks = await chain.readAll();
  const decoded = decodeStoredBlock(transport, 1, crypto, syncTestMkHex);
  t.assertEq(decoded, jsonSortNoSpaces(blocks[1]), 'P3: round-trips jsonSortNoSpaces');
  // Compact separators only — no `": "` / `", "` (spaces inside encrypted
  // string values, e.g. "Passage 0", are legitimate and not separators).
  t.assertEq(decoded.includes(': ') || decoded.includes(', '), false, 'P3: no space separators in payload');
});

// ── CPSW-P4 ────────────────────────────────────────────────────────
await test('CPSW-P4: pushAll throws on an empty chain (no genesis)', async () => {
  const crypto = makeCrypto();
  const chain = new CommonplaceChain(crypto, new MemoryBackend(), syncTestMkHex);
  const svc = new CommonplacePushService({ crypto, transport: new FakeSyncTransport(), chain });

  await t.assertAsyncThrows(svc.pushAll(), 'P4: throws on empty chain');
});

// ── CPSW-P5 ────────────────────────────────────────────────────────
await test('CPSW-P5: pushAll throws when no master key is cached', async () => {
  const crypto = makeCrypto(null);
  const chain = new CommonplaceChain(crypto, new MemoryBackend(), null);
  const svc = new CommonplacePushService({ crypto, transport: new FakeSyncTransport(), chain });

  await t.assertAsyncThrows(svc.pushAll(), 'P5: throws without MK');
});

// ── CPSW-P6 ────────────────────────────────────────────────────────
await test('CPSW-P6: pushBlocks pushes an explicit block list at 0-based positions', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  const result = await svc.pushBlocks(await chain.readAll());

  t.assertEq(result.pushed, 2, 'P6: two pushed');
  t.assert(transport.store.has(commonplaceBlockPath(0)), 'P6: path 0');
  t.assert(transport.store.has(commonplaceBlockPath(1)), 'P6: path 1');
  t.assertEq(readHashIndex(transport).length, 2, 'P6: hash index length 2');
});

// ── CPSW-P7 ────────────────────────────────────────────────────────
await test('CPSW-P7: repeated pushAll is idempotent — same paths overwritten', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  const first = await svc.pushAll();
  const second = await svc.pushAll();

  t.assertEq(first.pushed, 2, 'P7: first pushed 2');
  t.assertEq(second.pushed, 2, 'P7: second pushed 2');
  t.assert(transport.store.has(commonplaceBlockPath(0)), 'P7: path 0');
  t.assert(transport.store.has(commonplaceBlockPath(1)), 'P7: path 1');
  t.assert(transport.store.has(COMMONPLACE_HASH_INDEX), 'P7: hash index present');
});

// ── CPSW-P8 ────────────────────────────────────────────────────────
await test('CPSW-P8: a failing block yields pushed + failedBlocks with failed index', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  transport.errorOnPushPath[commonplaceBlockPath(1)] = 500;
  const svc = new CommonplacePushService({ crypto, transport, chain });

  const result = await svc.pushAll();

  t.assertEq(result.pushed, 1, 'P8: genesis pushed');
  t.assert(result.failedBlocks.includes(1), 'P8: index 1 in failedBlocks');
  t.assert(transport.store.has(commonplaceBlockPath(0)), 'P8: block 0 present');
});

// ── CPSW-P9 ────────────────────────────────────────────────────────
await test('CPSW-P9: concurrent pushAll calls are serialized (single push pass)', async () => {
  const crypto = makeCrypto();
  const chain = await buildChain(crypto, { dayBlocks: 1 });
  const transport = new FakeSyncTransport();
  const svc = new CommonplacePushService({ crypto, transport, chain });

  const f1 = svc.pushAll();
  const f2 = svc.pushAll();
  const r1 = await f1;
  const r2 = await f2;

  t.assertEq(r1.pushed, 2, 'P9: r1 pushed 2');
  t.assertEq(r2.pushed, 2, 'P9: r2 pushed 2');
  t.assertEq(transport.pushCount, 3, 'P9: 2 blocks + 1 hash_index pushed exactly once');
});

t.summary('CommonplacePushService');
process.exit(t.failed > 0 ? 1 : 0);

File: phpoc-flutter/test/services/commonplace_push_service_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:phpoc_flutter/data/commonplace/commonplace_chain.dart';
import 'package:phpoc_flutter/data/ledger/helpers.dart' show jsonEncodeSortedNoSpaces;
import 'package:phpoc_flutter/services/commonplace_push_service.dart';

import 'commonplace_sync_test_support.dart';

/// CommonplacePushService tests — Group P (9 assertions).
///
/// Blueprint: docs/planning/flutter/COMMONPLACE_BOOK_SYNC_PHASE1.md
///
/// Phase 2 (RED): the not-yet-created `CommonplacePushService`
/// (`lib/services/commonplace_push_service.dart`) makes these fail to compile.

void main() {
  group('P: CommonplacePushService — push to R2', () {
    test('CPSY-P1: pushAll pushes every block to commonplace/blocks/NNNNNN.json',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      final result = await svc.pushAll();

      expect(result.success, isTrue);
      expect(result.blocksPushed, 2);
      expect(transport.store.containsKey(commonplaceBlockPath(0)), isTrue);
      expect(transport.store.containsKey(commonplaceBlockPath(1)), isTrue);

      final genesis = jsonDecode(crypto.deobfuscateBlob(
        transport.store[commonplaceBlockPath(0)]!,
        syncTestMkHex,
      )) as Map<String, dynamic>;
      expect(genesis['type'], 'commonplace_genesis');

      final day = jsonDecode(crypto.deobfuscateBlob(
        transport.store[commonplaceBlockPath(1)]!,
        syncTestMkHex,
      )) as Map<String, dynamic>;
      expect(day['type'], 'commonplace');
    });

    test('CPSY-P2: pushAll pushes a plaintext hash_index of block hashes in order',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      await svc.pushAll();

      final index =
          jsonDecode(utf8.decode(transport.store[commonplaceHashIndexPath]!))
              as List<dynamic>;
      final blocks = chain.readAll();
      expect(index.length, 2);
      expect(index[0], chain.getBlockHashFor(blocks[0]));
      expect(index[1], chain.getBlockHashFor(blocks[1]));
    });

    test(
        'CPSY-P3: pushed block payload is sorted space-free PHPSPEC JSON '
        'obfuscated with the MK', () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      await svc.pushAll();

      final block = chain.readAll()[1];
      final obfuscated = transport.store[commonplaceBlockPath(1)]!;
      final decoded = crypto.deobfuscateBlob(obfuscated, syncTestMkHex);
      expect(decoded, jsonEncodeSortedNoSpaces(block));
      expect(decoded.contains(' '), isFalse);
    });

    test('CPSY-P4: pushAll throws StateError on an empty chain', () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = CommonplaceChain(crypto: crypto, store: FakeCommonplaceStore());
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: FakeSyncTransport(),
        chain: chain,
      );

      expect(svc.pushAll(), throwsStateError);
    });

    test('CPSY-P5: pushAll throws StateError when no master key is cached',
        () async {
      final crypto = initCrypto(); // no MK
      final chain = CommonplaceChain(crypto: crypto, store: FakeCommonplaceStore());
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: FakeSyncTransport(),
        chain: chain,
      );

      // The MK guard runs before the empty-chain guard, so this throws even
      // though the chain is empty.
      expect(svc.pushAll(), throwsStateError);
    });

    test('CPSY-P6: pushBlocks pushes an explicit block list at 0-based positions',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      final result = await svc.pushBlocks(chain.readAll());

      expect(result.success, isTrue);
      expect(result.blocksPushed, 2);
      expect(transport.store.containsKey(commonplaceBlockPath(0)), isTrue);
      expect(transport.store.containsKey(commonplaceBlockPath(1)), isTrue);
      final index =
          jsonDecode(utf8.decode(transport.store[commonplaceHashIndexPath]!))
              as List<dynamic>;
      expect(index.length, 2);
    });

    test('CPSY-P7: repeated pushAll is idempotent — same paths overwritten',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      final first = await svc.pushAll();
      final second = await svc.pushAll();

      expect(first.success, isTrue);
      expect(second.success, isTrue);
      expect(transport.store.containsKey(commonplaceBlockPath(0)), isTrue);
      expect(transport.store.containsKey(commonplaceBlockPath(1)), isTrue);
      expect(transport.store.containsKey(commonplaceHashIndexPath), isTrue);
    });

    test('CPSY-P8: a failing block yields PushResult.failure with failedBlocks',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      transport.errorOnPushPath[commonplaceBlockPath(1)] = 500;
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      final result = await svc.pushAll();

      expect(result.success, isFalse);
      expect(result.failedBlocks, contains(1));
      expect(transport.store.containsKey(commonplaceBlockPath(0)), isTrue);
    });

    test('CPSY-P9: concurrent pushAll calls are serialized (single push pass)',
        () async {
      final crypto = initCrypto(mkHex: syncTestMkHex);
      final chain = buildChain(crypto, dayBlocks: 1);
      final transport = FakeSyncTransport();
      final svc = CommonplacePushService(
        crypto: crypto,
        transport: transport,
        chain: chain,
      );

      final f1 = svc.pushAll();
      final f2 = svc.pushAll();
      final r1 = await f1;
      final r2 = await f2;

      expect(r1.success, isTrue);
      expect(r2.success, isTrue);
      // 2 blocks + 1 hash_index pushed exactly once (not twice).
      expect(transport.pushCount, 3);
    });
  });
}
